crates/sparrow-transforms/src/transform_pipeline.rs

```diff
@@ -0,0 +1,174 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use error_stack::ResultExt;
+use itertools::Itertools;
+use parking_lot::Mutex;
+use sparrow_arrow::Batch;
+use sparrow_scheduler::{
+    Partition, Partitioned, Pipeline, PipelineSink, Queue, Scheduler, Sink, TaskRef,
+};
+
+use crate::transforms::Transform;
+use crate::Error;
+
+/// Runs a linear sequence of transforms as a pipeline.
+pub struct TransformPipeline {
+    /// The inputs this transform processes.
+    ///
+    /// For each partition:
+    ///   Hold a shared, mutex-protected input queue.
+    inputs: Partitioned<Arc<Mutex<InputBuffer>>>,
+    transforms: Vec<Box<dyn Transform>>,
+    /// Compute tasks to wake up as needed.
+    tasks: Partitioned<TaskRef>,
+    /// Sink for the down-stream computation.
+    sink: PipelineSink,
+}
+
+impl std::fmt::Debug for TransformPipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TransformPipeline")
+            .field(
+                "transforms",
+                &self.transforms.iter().map(|t| t.name()).format(","),
+            )
+            .finish()
+    }
+}
+
+#[derive(Debug, Default)]
+struct InputBuffer {
+    buffer: VecDeque<Batch>,
+    is_closed: bool,
+}
+
+impl TransformPipeline {
+    pub(crate) fn schedule(
+        scheduler: &mut Scheduler,
+        partitions: usize,
+        transforms: Vec<Box<dyn Transform>>,
+        sink: PipelineSink,
+    ) {
+        scheduler.add_pipeline(partitions, |tasks| {
+            let input_partitions = tasks.len();
+            let mut inputs = Partitioned::with_capacity(input_partitions);
+            for _ in 0..input_partitions {
+                inputs.push(Arc::new(Mutex::new(InputBuffer::default())));
+            }
+            Self {
+                inputs,
+                transforms,
+                tasks,
+                sink,
+            }
+        })
+    }
+
+    fn pop_batch(&self, partition: Partition) -> Option<Batch> {
+        let mut input_partition = self.inputs[partition].lock();
+        input_partition.buffer.pop_front()
+    }
+
+    fn pending(&self, partition: Partition) -> usize {
+        self.inputs[partition].lock().buffer.len()
+    }
+
+    fn is_closed(&self, partition: Partition) -> bool {
+        self.inputs[partition].lock().is_closed
+    }
+}
+
+impl Pipeline for TransformPipeline {
+    type Context = Error;
+
+    fn push(
+        &self,
+        partition: Partition,
+        input: usize,
+        batch: Batch,
+        queue: &mut dyn Queue<TaskRef>,
+    ) -> error_stack::Result<(), Error> {
+        error_stack::ensure!(input == 0, Error::InvalidInput { input });
```

Comment:
What's this logic?

Reply:
Transform pipelines run on a single input, so it should always be 0.